### DISCUSSION TITLE
What is the correct way to create an OrtValue from a pytorch tensor on CUDA?

### DISCUSSION THREAD

**Discussion by mattroos** [microsoft/onnxruntime#7529] [Other Q&A] [answered]
In python, if I have a numpy array, ```x```,  and want to create an OrtValue on a CUDA device I do the following:
```ortvalue = onnxruntime.OrtValue.ortvalue_from_numpy(x, 'cuda', 0)```

But if I have a torch tensor, ```y```, that is already in CUDA/GPU memory, it seems I have to copy it to CPU memory before creating the OrtValue:
```onnxruntime.OrtValue.ortvalue_from_numpy(y.data.cpu().numpy(), 'cuda', 0)```

Is there a way to create the OrtValue directly from the the torch tensor, such that the tensor doesn't have to be copied to CPU memory and then back to GPU memory?

**Comment by hariharans29** [comment, marked as answer]
Hmm - I don't think we currently have a way to do this. What you want is to be able to provide a data pointer and metadata for that dat pointer: type, shape, and device information.

May I ask what your scenario is ? Do you want to feed an output from Torch as input to an ONNX graph via ORT's python interface and thereby looking to reduce the round-trip latency for the data ?

**Reply by mattroos** [reply]
> Do you want to feed an output from Torch as input to an ONNX graph via ORT's python interface and thereby looking to reduce the round-trip latency for the data?

Yes, that's correct. Down the road I may be able to rewrite a lot of my code to avoid using Torch, and just use numpy and the ORT, but for now I'm sticking with the torch code. I'm not sure what the latency savings would be by going directly from the torch cuda tensor to the OrtValue, but certainly would want to do that if it were an option.

**Reply by devymex** [reply]
I want to convert a torch tensor (on GPU) to OrtValues directly without via NumPy.

For my project retrieving data from a GPU decoder, I need a way to prevent the CPU-GPU copy.

**Reply by hariharans29** [reply]
@mattroos and @devymex - This is definitely a useful feature, please feel free to open a GitHub issue for this request

**Reply by jeshels** [reply]
Opened #10217

**Comment by nguyenvanhoang7398** [comment]
On the other hand, is there a way to convert OrtValue to Pytorch tensor on the same device? I need to run some tensor operations provided by PyTorch before passing the OrtValue to the next module.

**Reply by hariharans29** [reply]
@nguyenvanhoang7398 : The `OrtValue` instance does provide access to its underlying data buffer: https://github.com/microsoft/onnxruntime/blob/69ab4670f753980aa751c0032d1ac457d2d57a45/onnxruntime/python/onnxruntime_inference_collection.py#L560. Is it possible to wrap this buffer into a PyTorch tensor ?

**Comment by sradmard** [comment]
Has there been any progress made on this issue yet?

**Reply by sradmard** [reply]
@hariharans29 I was wondering whether there has been any progress on this issue or not. I would really appreciate it.

**Comment by farzanehnakhaee70** [comment]
Hi all,
I would also appreciated if there is any solution for this problem. I exactly have the same issue.

**Reply by jeshels** [reply]
https://github.com/microsoft/onnxruntime/issues/10217
